src/app/archive_hardfork_toolbox/sql.ml

```diff
@@ -0,0 +1,196 @@
+let number_of_internal_commands_since_block_query =
+  (Caqti_type.(t2 string int) ->! Caqti_type.(t4 string int int int))
+    {sql|
+    WITH RECURSIVE chain AS
+    (
+        SELECT id, parent_id, chain_status, state_hash, height, global_slot_since_genesis FROM blocks
+        WHERE state_hash = ?
+        UNION ALL
+        SELECT b.id, b.parent_id, b.chain_status, b.state_hash, b.height, b.global_slot_since_genesis FROM blocks b
+        INNER JOIN chain ON b.id = chain.parent_id AND (chain.id <> 0 OR b.id = 0)
+     ) SELECT state_hash,  height, global_slot_since_genesis, count(bc.block_id) as internal_command_count  FROM chain left join blocks_internal_commands bc on bc.block_id = id where global_slot_since_genesis >= ? group by state_hash, height, global_slot_since_genesis
+;
+    |sql}
+
+let number_of_internal_commands_since_block (module Conn : CONNECTION)
+    ~fork_state_hash ~fork_slot =
+  Conn.find number_of_internal_commands_since_block_query
+    (fork_state_hash, fork_slot)
+
+let number_of_zkapps_commands_since_block_query =
+  (Caqti_type.(t2 string int) ->! Caqti_type.(t4 string int int int))
+    {sql|
+    WITH RECURSIVE chain AS
+    (
+        SELECT id, parent_id, chain_status, state_hash, height, global_slot_since_genesis FROM blocks
+        WHERE state_hash = ?
+        UNION ALL
+        SELECT b.id, b.parent_id, b.chain_status, b.state_hash, b.height, b.global_slot_since_genesis FROM blocks b
+        INNER JOIN chain ON b.id = chain.parent_id AND (chain.id <> 0 OR b.id = 0)
+     ) SELECT state_hash,  height, global_slot_since_genesis, count(bc.block_id) as zkapp_command_count  FROM chain left join blocks_zkapp_commands bc on bc.block_id = id where global_slot_since_genesis >= ? group by state_hash, height, global_slot_since_genesis
+;
+    |sql}
+
+let number_of_zkapps_commands_since_block (module Conn : CONNECTION)
+    ~fork_state_hash ~fork_slot =
+  Conn.find number_of_zkapps_commands_since_block_query
+    (fork_state_hash, fork_slot)
+
+let last_fork_block_query =
+  (Caqti_type.unit ->! Caqti_type.(t2 string int64))
+    {sql|
+    SELECT state_hash, global_slot_since_genesis FROM blocks
+    WHERE global_slot_since_hard_fork = 0
+    ORDER BY height DESC
+    LIMIT 1;
+    |sql}
+
+let last_fork_block (module Conn : CONNECTION) =
+  Conn.find last_fork_block_query ()
+
+module SchemaVerification = struct
+  module Types = struct
+    type schema_row =
+      { status : string option
+      ; description : string option
+      ; applied_at : string
+      ; validated_at : string option
+      }
+
+    type result =
+      { missing_cols : int
+      ; total_fks : int
+      ; valid_fks : int
+      ; expected_cols_min : int
+      ; expected_cols_max : int
+      ; expected_fk_count : int
+      ; schema : schema_row option
+      ; ok_cols : bool
+      ; ok_fk_present : bool
+      ; ok_fk_validated : bool
+      ; ok_schema_status : bool
+      ; ok : bool
+      }
+  end
+
+  module Queries = struct
```

Comment:
Again, we've already relied on postgres to ensure the integrity of DB. not sure why are these check necessary.

Whenever a postgres script returns 0 that means the upgrade succeeds.